model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
+ *
+ *
+ */
+
+namespace oat\taoQtiItem\model\qti\CustomInteractionAsset;
+
+use oat\taoQtiItem\model\qti\CustomInteractionAsset\Extractor\BaseExtendedCustomInteractionAssetExtractor;
+use oat\taoQtiItem\model\qti\CustomInteractionAsset\Extractor\TextReaderExtendedAssetExtractor;
+use oat\taoQtiItem\model\qti\interaction\CustomInteraction;
+use oat\taoQtiItem\model\qti\interaction\ImsPortableCustomInteraction;
+use oat\taoQtiItem\model\qti\interaction\PortableCustomInteraction;
+
+/**
+ * @author Kiryl Poyu <kyril.poyu@taotesting.com>
+ */
+class ExtendedCustomInteractionAssetExtractorAllocator
+{
+    public const TEXT_READER_INTERACTION = 'textReaderInteraction';
+
+    public static function allocateExtractor(
+        CustomInteraction $interaction
+    ): ?BaseExtendedCustomInteractionAssetExtractor {
+        if (
+            $interaction instanceof ImsPortableCustomInteraction
+            || $interaction instanceof PortableCustomInteraction
+        ) {
+            if ($interaction->getTypeIdentifier() === self::TEXT_READER_INTERACTION) {
+                return new TextReaderExtendedAssetExtractor($interaction);
+            }
+        }
+
+        return null;
```

Comment:
Instead of a `null`. WDYT of returning a no-op implementation of the interface?

Reply:
probably we can, but i don't see any real profit from this. We just remove 1 `if` but create empty class which should return empty array for this iteration (IMO it looks strange)
```php
if ($assetExtractor !== null) {
            $extractedAssets = $assetExtractor->extract();
            foreach ($extractedAssets as $asset) {
                $this->addAsset('pci-resource', $asset);
            }
        }
```

Reply:
That's the point of a no-op service. Yes, you have one `if` now, but you may as well get some other at some other place.
null-objects instead of `null` values allow reducing the complexity of the code.

I won't be pushing for that change any more, but suggest you think about it a bit :)

Reply:
I agree, in case of repeated use there will be more `if`. You pushed it through :)